src/librustc_passes/check_const.rs

```diff
@@ -0,0 +1,136 @@
+        check_mod_const_bodies,
+        ..*providers
+    };
+}
+
+#[derive(Copy, Clone)]
+struct CheckConstVisitor<'tcx> {
+    sess: &'tcx Session,
+    hir_map: &'tcx Map<'tcx>,
+    kind: hir::BodyOwnerKind,
+}
+
+impl<'tcx> CheckConstVisitor<'tcx> {
+    fn new(tcx: TyCtxt<'tcx>) -> Self {
+        CheckConstVisitor {
+            sess: &tcx.sess,
+            hir_map: tcx.hir(),
+
+            // The initial value shouldn't matter since we need to update it during visitation.
+            // `Closure` was chosen because it will cause an error in `const_check_violated` if we
+            // ever forget to do that.
+            kind: hir::BodyOwnerKind::Closure,
+        }
+    }
+
+    fn visit_const_body(&mut self, kind: hir::BodyOwnerKind, body: &'tcx hir::Body) {
+        self.kind = kind;
+        self.visit_expr(&body.value);
+    }
+
+    /// Emits an error when an unsupported statement is found in a const item.
+    fn const_check_violated(&self, bad_op: &str, span: Span) {
+        use hir::BodyOwnerKind::*;
+
+        let const_kind = match self.kind {
+            Const => "const",
+            Static(_) => "static",
+            Fn => "const fn",
+            Closure => bug!("closures should never be const-checked"),
+        };
+
+        span_err!(self.sess, span, E0744, "{} is not allowed in a {}", bad_op, const_kind);
+    }
+}
+
+impl<'tcx> Visitor<'tcx> for CheckConstVisitor<'tcx> {
+    fn nested_visit_map<'this>(&'this mut self) -> NestedVisitorMap<'this, 'tcx> {
+        NestedVisitorMap::OnlyBodies(&self.hir_map)
+    }
+
+    fn visit_body(&mut self, body: &'tcx hir::Body) {
+        let owner = self.hir_map.body_owner(body.id());
+        let owner_kind = self.hir_map.body_owner_kind(owner);
+        match owner_kind {
+            | hir::BodyOwnerKind::Const
+            | hir::BodyOwnerKind::Static(_)
+            => self.visit_const_body(owner_kind, body),
+
+            hir::BodyOwnerKind::Fn => {
+                let fn_header = self.hir_map.fn_header_by_hir_id(owner).unwrap();
+                if fn_header.is_const() {
+                    self.visit_const_body(owner_kind, body);
+                }
+            }
+
+            hir::BodyOwnerKind::Closure => {}
+        }
+    }
+
+    fn visit_expr(&mut self, e: &'tcx hir::Expr) {
+        match &e.kind {
+            hir::ExprKind::Loop(_, _, source) => {
+                use hir::LoopSource::*;
+
+                let op = match source {
+                    Loop => "loop",
+                    While | WhileLet => "while loop",
+                    ForLoop => "for loop",
+                };
+
+                self.const_check_violated(op, e.span);
+            }
+
+            hir::ExprKind::Match(_, _, source) => {
+                use hir::MatchSource::*;
+
+                let op = match source {
+                    Normal => Some("match expression"),
+                    IfDesugar { .. } | IfLetDesugar { .. } => Some("if expression"),
```

Comment:
```suggestion
                    IfDesugar { .. } | IfLetDesugar { .. } => Some("`if` expression"),
```

Reply:
If we can do quotes, can we change this to e.g. "`if` is not allowed in a X"? The expression part makes it read clunkily to me.

Reply:
Sure!